extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/RetryableS3OutputStream.java

```diff
@@ -199,15 +221,44 @@ private void pushCurrentChunk() throws IOException
   {
     currentChunk.close();
     final Chunk chunk = currentChunk;
-    try {
-      if (chunk.length() > 0) {
-        resultsSize += chunk.length();
+    if (chunk.length() > 0) {
+      uploadManager.incrementCurrentNumChunks();
+      pendingFiles.incrementAndGet();
 
-        pushStopwatch.start();
-        pushResults.add(push(chunk));
-        pushStopwatch.stop();
-        numChunksPushed++;
+      uploadManager.submitTask(() -> {
+        try {
+          uploadChunk(chunk);
+        }
+        catch (Exception e) {
+          error = true;
+          LOG.error(e, e.getMessage());
+          throw new RuntimeException(e);
+        }
+        finally {
+          synchronized (maxChunksLock) {
+            uploadManager.decrementCurrentNumChunks();
+            maxChunksLock.notifyAll();
+          }
+          if (pendingFiles.decrementAndGet() == 0) {
+            synchronized (fileLock) {
+              fileLock.notifyAll();
+            }
+          }
+        }
+      });
+    }
+  }
```

Comment:
All the upload related logic should live in `S3UploadManager` itself, otherwise it is just an `ExecutorService`. Let me know if you need any assistance with this.

Reply:
@kfaraz I thought about that, but we have a couple of lock usages in RetryableS3OutputStream which can't be moved to S3UploadManager:
1. `maxChunksLock.wait();` in `write()` method
2. `fileLock.wait();` in `completeMultipartUpload()` method


So would you suggest I move these locks to `S3UploadManager` and expose methods like `getMaxChunksLock()` and `getFileLock()` to replace the current usages in `RetryableS3OutputStream`?

Reply:
Or maybe exposing a method like this in S3UploadManager?
```java
public boolean canWriteMoreChunksOnDisk()
  {
    synchronized (maxChunksLock) {
      while (uploadManager.getCurrentNumChunksOnDisk() > uploadManager.getMaxConcurrentNumChunks()) {
        try {
          LOG.debug("Waiting for lock for writing further chunks to local disk.");
          maxChunksLock.wait();
        }
        catch (InterruptedException e) {
          Thread.currentThread().interrupt();
          throw new RuntimeException(e);
        }
      }
    }
  }
```

And replacing the above code in RetryableS3OutputStream with `canWriteMoreChunksOnDisk()` 🤔

Reply:
@kfaraz I tried sketching out how it could look, and I don't think it'll be clean to move the upload functionality to outside of RetryableS3OutputStream file.

RetryableS3OutputStream is a per-`uploadId` class. Most/All of its fields are per upload ID info.

To move the upload functionality to a separate Singleton class like `S3UploadManager`, we would need a `ConcurrentHashMap: uploadId -> xyz` in the Singleton for every field like `pendingFiles`, `fileLock`, `resultsSize`, `numChunksPushed`, `pushStopwatch`, `pushResults`, `error`, `closed`, etc - which seems like an overkill to me.

We'd also have to pass `S3OutputConfig`, `s3Key` and the s3 client to the Singleton class, that is, to S3UploadManager.

Thoughts? Maybe we can rename `S3UploadManager` to something else indicating that it doesn't do the actual upload? The original intention was to keep and manage only the global (that is, non-uploadId-specific) components in the Singleton class, as RetryableS3OutputStream would've needed static fields to achieve it otherwise.

I'll wait for your thoughts before tinkering further on this. Thanks!